experimental/devices/mfrc522/mfrc522.go

```diff
@@ -20,107 +20,81 @@ import (
-	RAB_WB_IB_DAB             = 0x06
-	RAB_WN_IN_DAB             = 0x01
-	RB_WB_IN_DN               = 0x03
-	RB_WN_IN_DN               = 0x05
-	RN_WN_IN_DN               = 0x07
-
-	KeyA_RN_WA_BITS_RA_WN_KeyB_RA_WA        SectorTrailerAccess = iota
-	KeyA_RN_WN_BITS_RA_WN_KeyB_RA_WN                            = 0x02
-	KeyA_RN_WB_BITS_RAB_WN_KeyB_RN_WB                           = 0x04
-	KeyA_RN_WN_BITS_RAB_WN_KeyB_RN_WN                           = 0x06
-	KeyA_RN_WA_BITS_RA_WA_KeyB_RA_WA                            = 0x01
-	KeyA_RN_WB_BITS_RAB_WB_KeyB_RN_WB                           = 0x03
-	KeyA_RN_WN_BITS_RAB_WB_KeyB_RN_WN                           = 0x05
-	KeyA_RN_WN_BITS_RAB_WN_KeyB_RN_WN_EXTRA                     = 0x07
-)
-
-// BlocksAccess defines the access structure for first 3 blocks of the sector
-// and the access bits for the sector trail.
-type BlocksAccess struct {
-	B0, B1, B2 BlockAccess
-	B3         SectorTrailerAccess
+// Dev is an handle to an MFRC522 RFID reader.
+type Dev struct {
+	LowLevel         *commands.LowLevel
+	operationTimeout time.Duration
+	beforeCall       func()
+	afterCall        func()
 }
 
-func (ba *BlocksAccess) String() string {
-	return fmt.Sprintf("B0: %d, B1: %d, B2: %d, B3: %d", ba.B0, ba.B1, ba.B2, ba.B3)
+// Key is the access key that consists of 6 bytes. There could be two types of keys - keyA and keyB.
+// KeyA and KeyB correspond to the different sector trail and data access. Refer to the datasheet for more details.
+type Key [6]byte
+
+// DefaultKey  provides the default bytes for card authentication for method B.
+var DefaultKey = Key{0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}
+
+type config struct {
+	defaultTimeout time.Duration
+	beforeCall     func()
+	afterCall      func()
 }
 
-// CalculateBlockAccess calculates the block access.
-func CalculateBlockAccess(ba *BlocksAccess) []byte {
-	res := make([]byte, 4)
-	res[0] = ((^ba.getBits(2) & 0x0F) << 4) | (^ba.getBits(1) & 0x0F)
-	res[1] = ((ba.getBits(1) & 0x0F) << 4) | (^ba.getBits(3) & 0x0F)
-	res[2] = ((ba.getBits(3) & 0x0F) << 4) | (ba.getBits(2) & 0x0F)
-	res[3] = res[0] ^ res[1] ^ res[2]
-	return res
+type configF func(*config) *config
+
+// WithTimeout updates the default device-wide configuration timeout.
+func WithTimeout(timeout time.Duration) configF {
+	return func(c *config) *config {
+		c.defaultTimeout = timeout
+		return c
+	}
 }
 
-// ParseBlockAccess parses the given byte array into the block access structure.
-func ParseBlockAccess(ad []byte) *BlocksAccess {
-	return &BlocksAccess{
-		B0: BlockAccess(((ad[1] & 0x10) >> 2) | ((ad[2] & 0x01) << 1) | ((ad[2] & 0x10) >> 5)),
-		B1: BlockAccess(((ad[1] & 0x20) >> 3) | (ad[2] & 0x02) | ((ad[2] & 0x20) >> 5)),
-		B2: BlockAccess(((ad[1] & 0x40) >> 4) | ((ad[2] & 0x04) >> 1) | ((ad[2] & 0x40) >> 6)),
-		B3: SectorTrailerAccess(((ad[1] & 0x80) >> 5) | ((ad[2] & 0x08) >> 2) | ((ad[2] & 0x80) >> 7)),
+// WithSyncsets the synchronization for the entire device, using internal mutex.
```

Comment:
That is the way to add the sync over the device as a "trait". If someone wants to use that device with the "locks" - then with this configuration wrapper for before/after all the operations will be synchronized using the private mutex. 
Otherwise for a single-threaded access no mutex is necessary. Hence no need for an embedded mutex at all.